sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/consumer/strategies/retryprocessingerror/RetryProcessingErrorMLS.java

```diff
@@ -105,4 +124,46 @@ public void verifyConsumerConfig(Map<String, String> config) {
           "The strategy should commit explicitly by partition but property 'enable.auto.commit' in consumer config is set to 'true'");
     }
   }
+
+  private void callErrorHandler(
+      ErrorHandler<K, V> errorHandler,
+      ConsumerRecord<K, V> consumerRecord,
+      RuntimeException e,
+      KafkaConsumer<K, V> consumer) {
+    if (errorHandler != null) {
+      boolean shouldContinue = errorHandler.handleError(consumerRecord, e, consumer);
+      if (!shouldContinue) {
+        throw new StopListenerException(e);
+      }
+    }
+  }
+
+  private OffsetAndMetadata markConsumerRecordProcessed(ConsumerRecord<K, V> consumerRecord) {
+    addOffsetToCommitOnClose(consumerRecord);
+    return new OffsetAndMetadata(consumerRecord.offset() + 1);
+  }
+
+  private OffsetCounter getOffsetCounter(TopicPartition partition, long offset) {
+    OffsetCounter counter = offsetCounters.computeIfAbsent(partition, k -> new OffsetCounter(0));
```

Comment:
My understanding: By tracking and comparing the offset, you effectively realize when a retried record is skipped after reaching the retry limit or processed successfully. I think for me it would be easier to understand, if we don't take care about the offset here and just remove the partition from the map on success or skip explicitly.

Reply:
When taking the offset into account, I can get the current count just out of the context, without the help of any other code.

The other way I have to rely on the code finishing message processing to reset the current count. I think this way it is more robust

Reply:
Valid point. I wonder, if it is just me who needs too much time, to understand what this code is doing. Maybe we can help there a bit, e.g. JavaDoc or inline comments for the method and/or the map, more descriptive method name (what about `getRetryCounterForOffset`?) and class name (what about `OffsetAwareRetryCounter`? The current name counts offsets in my mind).

Reply:
It may also be an option to extract the whole tracking (this logic and the map) and maybe the retry limit exceeded decision in a dedicated class with a clean API. But I have no idea yet, how that could look like.

Reply:
I refactred the count-logic to another class and improved comments. Let's see if that helps to make things more readable.
I didn't change the name to keep things backward compatible.